compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1446,7 +1447,36 @@ trait Applications extends Compatibility {
             unapplyArgType
 
         val dummyArg = dummyTreeOfType(ownType)
-        val unapplyApp = typedExpr(untpd.TypedSplice(Apply(unapplyFn, dummyArg :: Nil)))
+        val (newUnapplyFn, unapplyApp) =
+          val unapplyAppCall = withMode(Mode.NoInline) {
+            typedExpr(untpd.TypedSplice(Apply(unapplyFn, dummyArg :: Nil)))
+          }
+          /** Inlines the unapply function before the dummy argument
+           *
+           *  A call `P.unapply[...](using l1, ..)(`dummy`)(using t1, ..)` becomes
+           *  ```
+           *  {
+           *    class $anon {
+           *      def unapply(s: S)(using t1: T1, ..): R =
+           *        ... // inlined code for: P.unapply[...](using l1, ..)(s)(using t1, ..)
+           *    }
+           *    new $anon
+           *  }.unapply(`dummy`)(using t1, ..)
+           *  ```
+           */
+          def inlinedUnapplyFnAndApp(unapp: Tree): (Tree, Tree) = unapp match {
```

Comment:
Consider moving this function further out, into the body of `typedUnapply`. It's a complicated operation for a special case, which makes it harder to follow the already complicated logic here.

Reply:
Moved out `inlinedUnapplyFnAndApp`. Also moved `unapplyImplicits` in a separate commit to cleanup the code even further.